### PR TITLE
Make InputHandlers pass transient change information instead of full states

### LINKED ISSUE #1612: InputManager can trigger OnMouseMove with empty MouseState in some circumstances
I think this is a regression introduced by #1595.
All `CurrentState` fields are now always non-null:

https://github.com/ppy/osu-framework/blob/19c03161d4351f4f65a5860b6a2f0d37dbe129d4/osu.Framework/Input/InputManager.cs#L60-L69

But this check still exists:

https://github.com/ppy/osu-framework/blob/19c03161d4351f4f65a5860b6a2f0d37dbe129d4/osu.Framework/Input/InputManager.cs#L220-L226

This prevented `OnMouseMove` from being triggered when there was no `MouseState`. But now that this is always true, it is always triggered, even when the `MouseState` is empty, leading to unexpected results (e.g. setting cursor to [0, 0], see ppy/osu#2754)

It should be replaced by some kind of check which validates that the `CurrentState.Mouse` field is not empty, but I am not sure how to do it.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/59429870-inputmanager-can-trigger-onmousemove-with-empty-mousestate-in-some-circumstances?utm_campaign=plugin&utm_content=tracker%2F44024461&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44024461&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ekrctb**
Implementation:
- `InputHandler`s produce `IInput`s and be queued to `PendingInputs`. Those are then handled by `InputManager`. For an `IInput`, `CurrentState` is modified in-place and if there is a change then `HandleMousePositionChange` etc. will be called. The handlers do appropriate processing and events will be produced.
- `MousePositionAbsoluteInput` and `MousePositionRelativeInput` are `IInput`. The latter is used for raw mouse input and the former is used for anything else.
- `ButtonStates<TButton>` and `ButtonInput<TButton>` are used in order to implement common logics of mouse button, keyboard key and joystick button handling.

Goal:
- Remove the need of createDistinctStates and their code duplication.
- Properly distinguish relative input and absolute input. Mouse cursor position is either relative or absolute and mouse scroll is relative.

To do:
- Test with absolute mouse positioning, precise scrolling and joystick input (I don't have devices).

Side effects:
- PassThroughInputManager used to be delayed by a frame. Made it doesn't delay.
- API change affecting osu's FramedReplayInputHandler and RulesetInputManager.
  can be found [here](https://github.com/ekrctb/osu/tree/new-input-manager) (not merge-able to master for now).
- Joystick axes are unimplemented. It is currently unused and I think raw axis values are useless. I want to keep it unimplemented until some new design will be invented.
- When a device disconnected while a button is pressed, the game thinks the button is still pressed. It is planned that this bug will be addressed in non-boolean button state PR (see below).

Future work (not addressed in this PR):
- InputState's button states should not be a boolean but rather than a counter.
- InputState should't have deltas.
- InputState cloning should be avoided in general.
  I added some additional cloning too.
- Input queue optimization or eliminate the input queue thing at all.

Feedback for naming is welcome.

Should resolve ppy/osu#2235.
Fixes #1612.

---

This change allows input handlers (keyboard / mouse / joystick) to return only change information (ie. "Button1 was Pressed") rather than full states which are then post-processed in a central location. It tidies up input handling logic and allows for future optimisations.

Note that this change in itself is not an optimisation and may temporarily reduce performance (but seems to be imperceivable). It does fix a single frame delay that could be experienced during gameplay, though.

### CODE REVIEW

**Comment by ekrctb** [comment]
It should be working now.

File: osu.Framework/Input/ButtonInput.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        public ButtonInputEntry(TButton button, bool isPressed)
+        {
+            Button = button;
+            IsPressed = isPressed;
+        }
+    }
+
+    internal static class ButtonInputHelper
+    {
+        public static TInput MakeSingleton<TInput, TButton>(TButton button, bool isPressed)
```
Comment (peppy) [inline osu.Framework/Input/ButtonInput.cs]: i don't think `Singleton` is the correct term here. should probably just be `CreateInput`, or possibly just be a constructor in the first place? we always prefer ctors over static helper classes where possible.

**Comment by peppy** [comment]
I'm going through this in fine detail and will point out individual issues in a code review, but I'll mention that this needs xmldoc for all public classes/methods.

**Comment by peppy** [comment]
Focusing on the input queue building first of all, I took the liberty of merging two PRs I opened today to help further discussion/refactoring.

A few things about when queues need to be built/not build

- Mouse (positional) queue building and hover checks need to be run every frame, regardless of cursor movement. This is to cover the case where the user input has not changed but drawables have changed (causing hover changes). Note that this build/check can be skipped if there is any user input as an optimisation – we just need to be sure queue/hover is processed once or more.
- Keyboard queue building does not need to happen *at all* unless a key is down. It is more expensive to build than the mouse queue as there are no positional checks, so should definitely be avoided when not needed.
- Queues need to be reprocessed for each input event, possibly only in the case the event was blocked (see `TestCaseInputQueueChange` which I created to test against). It should be self-explanatory in why this is required.

**Comment by ekrctb** [comment]
I change the way the input queues built. The input queues are rebuilt just before when it is used.
In this way correctness is clear.
I want to leave optimizations to future PRs which can have more changes rather than a just suppress input queue rebuilds.

File: osu.Framework/Input/IMouseState.cs [hunk 2]
```diff
@@ -11,12 +10,14 @@ public interface IMouseState
...
         IMouseState NativeState { get; }
 
-        IReadOnlyList<MouseButton> Buttons { get; set; }
+        ButtonStates<MouseButton> Buttons { get; }
 
         Vector2 Delta { get; }
 
         Vector2 Position { get; set; }
 
+        bool IsPositionValid { get; set; }
```
Comment (peppy) [inline osu.Framework/Input/IMouseState.cs]: this should get some xmldoc to make its purpose clear. honestly this whole interface could do with xmldoc if you can be bothered adding it, but at very least we try to document new additions.

**Comment by smoogipoo** [comment]
Looks like I'm getting a very large scrollwheel delta when scrolling in the game after having scrolled outside: https://streamable.com/llj5c

**Comment by ekrctb** [comment]
The bug above is fixed.

**Comment by Tom94** [comment]
In general I am in favor of moving input handlers to provide more specialized state changes. I currently don't have the capacity to review a large change like this one, so feel free to merge! I trust your judgement. :)

**Pull request merged**